mmpy_bot/function.py

```diff
@@ -216,15 +188,20 @@ def wrapped_func(func):
             reg = f"^{reg.strip('^')} (.*)?"  # noqa
 
         pattern = re.compile(reg, regexp_flag)
-        return MessageFunction(
+        new_func = MessageFunction(
             func,
             matcher=pattern,
             direct_only=direct_only,
             needs_mention=needs_mention,
             allowed_users=allowed_users,
             allowed_channels=allowed_channels,
+            **annotations,
         )
 
+        # Preserve docstring
+        new_func.__doc__ = func.__doc__
```

Comment:
Why is this necessary? We can definitely explicitly store the `__doc__` of the wrapped function somewhere, but I'd prefer not to modify the `__doc__` of the wrapping `Function` since `__doc__` is normally class-specific, not instance-specific.

Reply:
This one was tricky. We now have `PluginHelpInfo.function_doc[header|full]` as well as `PluginHelpInfo.plugin_doc[header|full]` which contain the method and class level `Plugin` docstring, respectively.
With the way decorated functions are transformed into `MessageFunction` or `WebHookFunction` and there can be multiple decorators (multiple listen patterns, click decorators, ...), this was the simplest way to propagate the docstring without having to do some kind of conditional code using `isinstance()`.

The test `tests/unit_tests/plugin_manager_test.py::TestPluginManager::test_get_help` verifies if the docstring is propagated correctly. Commenting these lines breaks it.

Reply:
I still don't understand this. The base `Function` class has `self.docstring = self.function.__doc__ or ""`, where `self.function` is always the absolute innermost function (i.e. if you pass it a Function, it will keep unwrapping it until it reaches the callable that this Function wraps). Why then would we need to change `self.__doc__` as well, and can't we just use `self.docstring`?

Reply:
There might be some confusion on my part. When I tried to capture this in a test, the test fails unless I included the line above.

Can you perhaps have a look at the test `tests/unit_tests/plugin_manager_test.py::TestPluginManager::test_get_help` and confirm that the test and its assumptions are correct?

Reply:
I can't  judge what's going on just from looking at the code, so will have to try running it later